### ISSUE TITLE
Surprising behaviour of integers and subranges

### ISSUE THREAD

**Issue by markus-oberhumer** [nim-lang/Nim#5324]
This is a followup to issue #3434.

The Nim manual talks about [Type relations](https://nim-lang.org/docs/manual.html#type-relations), but I'm missing a concise definition of Nim's integer model, i.e. automatic integral promotion and conversion, and their interoperability with [Subranges](https://nim-lang.org/docs/manual.html#types-subrange-types).

This example shows the quite surprising behaviour of the current version (nim 0.16.1-git-137b5f4):

```
var data: array[4, float]

### uint

proc test_and_3(a: uint): float =
  let a = a and 3
  return data[a]

proc test_mod(a: uint): float =
  let a = a mod 4
  return data[a]

### uint16

proc test_access(a: uint16): float =
  return data[a]

proc test_add(a: uint16): float =
  let a = a + 1
  return data[a]

proc test_sub(a: uint16): float =
  let a = a - 1
  return data[a]

proc test_mul(a: uint16): float =
  let a = a * 2
  return data[a]

proc test_div(a: uint16): float =
  let a = a div 2
  return data[a]

proc test_or(a: uint16): float =
  let a = a or 2
  return data[a]

proc test_xor(a: uint16): float =
  let a = a xor 2
  return data[a]

proc test_and(a: uint16): float =
  let a = a and 2                              # works - this is NOT range 2..3(uint16)
  return data[a]

proc test_and_3(a: uint16): float =
  let a = a and 3                              # DOES NOT COMPILE !!
  return data[a]

proc test_mod(a: uint16): float =
  let a = a mod 4                              # DOES NOT COMPILE !!
  return data[a]
```

**Comment by linkerstorm** [comment]
Hi.

Perhaps a subrange bug related to the yours :
```nim
type Int_Subrange = range[1..5]
var x_Int_Subrange : Int_Subrange
echo x_Int_Subrange # Print zero, which is out of subrange

type Count_Enum = enum
  e1, e2, e3, e4, e5
type Count_Subrange = range[e2..e4]
var x_Count_Subrange : Count_Subrange
echo x_Count_Subrange # Print 'e1', which is out of subrange

var x_Positive : Positive
echo x_Positive # Print zero, which is out of subrange
```

**Comment by Araq** [comment]
> we should assume that the implicit initialization sets the first value of the subrange.

No instead the compiler warns you about this. This is also covered in the manual.

**Comment by markus-oberhumer** [comment]
This was fixed by a language change in git commit d7a896f19dfc7ff2ad22b78ef03e4db73332fc7e.
